storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,24 +3580,347 @@ dberr_t fsp_sys_tablespace_validate()
+      if (mach_read_from_8(FSEG_ID + inode) == 0)
+        continue;
+      if (!inodes->find(addr.page, offset))
+      {
+        err= fseg_inode_free(block, inode, space, mtr);
+        if (err)
+          return err;
+      }
+    }
+    addr= next_addr;
+  }
+  ut_ad(addr.page == FIL_NULL);
+  return err;
+}
+
+/** Remove the unused segment in tablespace. This function
+used only during shrinking of system tablespace
+@param mtr       mini-transaction
+@param shutdown  Called during shutdown
+@return error code */
+dberr_t fil_space_t::garbage_collect(mtr_t *mtr, bool shutdown)
+{
+  if (!shutdown && !trx_sys.is_undo_empty())
+  {
+    sql_print_warning("InnoDB: Can't free the unused segments"
+                      " in system tablespace because a previous shutdown"
+                      " was not with innodb_fast_shutdown=0");
+    return DB_ERROR;
+  }
+
+  ut_a(id == 0);
+  inode_info inodes;
+  dberr_t err= fsp_get_sys_used_segment(&inodes, mtr);
+  if (err)
+    return err;
+
+  mtr->x_lock_space(this);
+  err= fsp_free_unused_seg(this, mtr,
+                           FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL,
+                           &inodes);
+  if (err)
+    return err;
+  err= fsp_free_unused_seg(this, mtr,
+                           FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE,
+                           &inodes);
+  if (err)
+    return err;
+
+  /* Reset rollback segment header page slots because if we upgrade
+  from older version, there is a possiblity that UNDO_CACHED page
+  exist in rollback segment page slots for reuse purpose */
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS; rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == id)
+    {
+      buf_block_t *block= buf_page_get_gen(rseg->page_id(), 0,
+                                           RW_X_LATCH, nullptr,
+                                           BUF_GET, mtr, &err);
+      if (!block)
+        return err;
+      mtr->memset(block,
+                  TRX_RSEG + TRX_RSEG_UNDO_SLOTS,
+                  TRX_RSEG_N_SLOTS * TRX_RSEG_SLOT_SIZE, 0xff);
+    }
+  }
+  return err;
+}
+
+void fsp_system_tablespace_truncate(bool shutdown)
 {
   ut_ad(!purge_sys.enabled());
   ut_ad(!srv_undo_sources);
   uint32_t last_used_extent= 0;
   fil_space_t *space= fil_system.sys_space;
   mtr_t mtr;
   mtr.start();
-  mtr.x_lock_space(space);
-  dberr_t err= fsp_traverse_extents(space, &last_used_extent, &mtr);
-  if (err != DB_SUCCESS)
+  dberr_t err= space->garbage_collect(&mtr, shutdown);
+  if (err)
   {
 func_exit:
-    sql_print_warning("InnoDB: Cannot shrink the system tablespace "
-                      "due to %s", ut_strerr(err));
+    sql_print_warning(
+      "InnoDB: Cannot shrink the system tablespace due to %s",
+      ut_strerr(err));
     mtr.commit();
     return;
   }
+
+  err= fsp_traverse_extents(space, &last_used_extent, &mtr);
+  if (err != DB_SUCCESS)
+    goto func_exit;
```

Comment:
`space->garbage_collect()` already changed some things. Can we invoke `mtr.commit()` and `mtr.start()` before this step? This is about actually shrinking the system tablespace file size, right? If it is about something else, then could we do it as part of `garbage_collect()`? The goal should be to first ensure that the operation is not going to fail, before starting to modify things. There is no rollback, after all.